Replace symlinks with copyfile for cross-platform compatibility

jaxom-widby-nttd opened the pull request:
Update manifests to replace symlinks with copyfile, enhancing compatibility across different platforms. Additionally, add a pre-commit configuration file and update the .tool-versions for various tools.

jaxom-widby-nttd committed:
feat: replace symlinks with copyfile in manifests for better cross-platform compatibility

jaxom-widby-nttd committed:
feat: add forbidden files check and update secrets baseline in pre-commit config; Also added .secrets.baseline reference into the manifest.xml for terraform module items as required in .pre-commit-config.yaml

chris11-taylor-nttd commented on .pre-commit-config.yaml:
```diff
@@ -0,0 +1,61 @@
+  #       entry: golangci-lint run --fix
+  #       types: [go]
+  #       language: golang
+  #       pass_filenames: false
```
```suggestion
```

Let's just remove these commented sections, the original intent was that the component repositories for each type (terraform, python, etc.) would provide their own pre-commit configuration, so there's no harm in removing these unused ones.

chris11-taylor-nttd commented on examples/.tool-versions:
Probably worth including one of the tools that has a custom installation in this example file:

https://github.com/launchbynttdata/tf-azurerm-module_primitive-web_app/blob/main/.tool-versions#L5

We've customized the behavior so that if a tool and version are followed by a comment with a URL, we use that URL to add the plugin to the local asdf. Eventually I'd like to see this move to `mise` which has better support through the TOML format for this sort of thing, but for now it's probably best we have it in the example here so that we can test it.

jaxom-widby-nttd replied:
Ah Gotcha, I had been carrying that forward on recent modules and was wondering why regula was giving me issues (other than its a little long in the tooth).

Also mentioned that over into: https://github.com/launchbynttdata/tf-aws-module-template/pull/1/files#r2633149964

Line in particular to update would be to move to ```regula 3.2.1 # https://github.com/launchbynttdata/asdf-regula```

---

Will make that change here in a minute...

jaxom-widby-nttd replied:
Resolved in https://github.com/launchbynttdata/launch-common-automation-framework/pull/37/commits/3f4c73f9d329e1fb9760555ae5467ae66d5aab29

chris11-taylor-nttd commented on examples/Makefile:
With the way this Makefile is currently configured, it's going to try to refer back to a very old version of this repo (1.0.0). I would suggest we either bump the REPO_MANIFESTS_REVISION here to the tag you intend to release with this update, or maybe we just set this one to track `main` rather than a tag.

jaxom-widby-nttd replied:
In  https://github.com/launchbynttdata/launch-common-automation-framework/pull/37/commits/45f30a9855d45d9e4ee0119619ec32e5f494a5b0 I updated the Makefile referenced here to the one we have been utilizing on many of the recent modules overall that @ben-vaughan-nttd has proposed over in https://github.com/launchbynttdata/tf-aws-module-template/blob/8fe401a2067d37c3198c50e7f8aecfa6a6a33800/Makefile

---

Carried that over and committed. We may want to entertain keeping a tag like "prod" that might reduce the chore potential.

chris11-taylor-nttd reviewed (dismissed):
I'm on board with using copyfile over linkfile assuming nothing comes up in testing.

Left some thoughts on a few files but overall looks good.

jaxom-widby-nttd committed:
Update .pre-commit-config.yaml

Makes sense to me, mostly left it there just while testing ahead of clean up

Co-authored-by: Chris Taylor <132399041+chris11-taylor-nttd@users.noreply.github.com>
Signed-off-by: jaxom-widby-nttd <jaxom.widby@nttdata.com>

jaxom-widby-nttd committed:
re-add regula with URL fallback

Signed-off-by: jaxom-widby-nttd <jaxom.widby@nttdata.com>

jaxom-widby-nttd committed:
Refactor Makefile for repository configuration to use updated standard

Since we were planning on updating the version in one of the tags referenced here I took the opportunity to carry forward some of the other changes that are present within the "working" Makefile proposed/used commonly and stored in: https://github.com/launchbynttdata/tf-aws-module-template/blob/8fe401a2067d37c3198c50e7f8aecfa6a6a33800/Makefile

Signed-off-by: jaxom-widby-nttd <jaxom.widby@nttdata.com>

jaxom-widby-nttd commented on manifests/terraform_modules/manifest/manifest.xml:
```diff
@@ -1,14 +1,16 @@
+    <copyfile src="linkfiles/.tflint.hcl" dest=".tflint.hcl" />
+    <copyfile src="linkfiles/.pre-commit-config.yaml" dest=".pre-commit-config.yaml" />
+    <copyfile src="linkfiles/.golangci.yaml" dest=".golangci.yaml" />
+    <copyfile src="linkfiles/.secrets.baseline" dest=".secrets.baseline" />
```
This requires that [.secrets.baseline](https://github.com/launchbynttdata/lcaf-component-terraform/blob/2578386f7b6c120aad73e0723259495b2933bac8/linkfiles/.secrets.baseline) in https://github.com/launchbynttdata/lcaf-component-terraform/pull/18 be committed first before this is rolled out.

ben-vaughan-nttd replied:
We have been including these files directly in the repos, rather than relying on LCAF to copy them in. IMO, we should remove them entirely from being managed by this manifest. Is there a way to configure this to only copy the file if they are not already present?

Additionally, I think the `.secrets.baseline` should not be included in the skeleton. We should either create a target in the Makefile to help us manage this file directly with the detect-secrets tool, or something similar in the pre-commit configuration.

jaxom-widby-nttd replied:
I agree was just adding the updates as are deployed in this. while an interesting method the Makefile should be easily able to handle this same activity in a nicer (imo) manner. Though depends on the roadmap of where this is moving towards. At the moment its partially a chicken or egg scenario as if the pre-commit-config references the secrets.baseline and its not present it throws an error to resolve each time. For me its become part of the standard workflow of carrying forward partially seeded template repo files instead of having the template do the deployment part.

jaxom-widby-nttd replied:
Updating this to remove the .secrets.baseline reference as the repo maintainer/author should be running "detect-secrets scan > .secrets.baseline" during the process ahead of commits to make the pre-commit run happy.

(removed line 13)

Leaving this conversation open to review with @chris11-taylor-nttd and @ben-vaughan-nttd  at least on the potential for removing more from this manifest instead of updating to copyfile or how might this best be handled.

jaxom-widby-nttd committed:
Remove .secrets.baseline copyfile from manifest

Removed copyfile entry for .secrets.baseline as author/maintainer should be running with "detect-secrets scan > .secrets.baseline" at key touchpoints to create the file.

Signed-off-by: jaxom-widby-nttd <jaxom.widby@nttdata.com>
